### PR TITLE
Biologically-catalysed reactions in Geochemistry

### PR DESCRIPTION
**Pull request by WilkAndy**
moose_geochemdb.json modified by adding some Biomass species and Mackinawite mineral in preparation for biogeochem.

Refs #21559

### CODE REVIEW

**Comment by WilkAndy** [comment]
@loganharbour , i'm hoping you'll review this as i go.  My hope is to get this all done within a month.  TODO:

- [ ] core code changes
- [ ] unit-test code changes so existing unit tests run with the core code changes
- [ ] nonunit-test changes so the existing tests run with the core code changes
- [ ] documentation of the core code changes
- [ ] unit tests that comprehensively test the core code changes
- [ ] nonunit tests that demonstrate some of the new features
- [ ] documentation of biogeochemistry
- [ ] fully-fledged example(s?) of batch biogeochemistry
- [ ] fully-fledged example(s?) of reactive transport with biogeochemistry

**Comment by WilkAndy** [comment]
Hi @loganharbour - this is ready for review!

**Comment by WilkAndy** [comment]
Hi @loganharbour - just checking this is on your "to-do" list and you haven't silenced notifications from github.  There's no rush on this one - i just want to know whether to email you!

**Comment by WilkAndy** [comment]
@loganharbour - i just remembered about this!!   Is it possible to review it in the next fortnight?

**Comment by loganharbour** [comment]
Looks good to me so far. We really need to move images eventually over to the large_media submodule - maybe this is a good time for it? I've got a few files left. Still haven't looked at documentation. Is there someone you want to bother for the doco?

**Comment by WilkAndy** [comment]
Hi @cpgr , would you mind reviewing the documentation for this PR?

**Comment by cpgr** [comment]
> Hi @cpgr , would you mind reviewing the documentation for this PR?

Will do it today sometime

File: modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_zoning.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+
+
+## Results
+
+As seen in [bio_zoning.mov], the simulation gradual evolves to steady-state in which the aquifer is partitioned into two distinct zones: the upstream zone is inhabited by the sulfate reducers, who drive the sulfate concentration to low values by around 100$\,$km from the fresh groundwater source (which sits at $x=0$).  The downstream zone is inhabited by methanogens, who produce significant dissolved methane.
+
+!media geochemistry/bio_zoning.mp4 caption=Results of the aquifer-zoning biogeochemistry reactive transport simulation  id=bio_zoning.mov
+
+
```
Comment (cpgr) [inline modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_zoning.md:147]: This looks so cool! I'm really impressed that you have added this capability!

**Comment by WilkAndy** [comment]
Thank you @cpgr

File: modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# Simulating biogeochemistry
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:1]: done

File: modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Studying the above database entry reveals the following.
+
+1. The reactants in the original [methanogen.cata] (CH$_{3}$COO$^{-}$ in this case) must be brought to the right-hand side of the equation.  This has the interpretation that "when a methanogen reacts, it removes acetate and water to produce methane and biocarbonate", viz:
+
+\begin{equation}
+\mathrm{methanogen} \rightleftharpoons -\mathrm{CH}_{3}\mathrm{COO}^{-} + \mathrm{CH}_{4}\mathrm{(aq)} + \mathrm{HCO}_{3}^{-} - \mathrm{H}_{2}\mathrm{O} \ .
+\end{equation}
+
+2. The `molar volume` and `molecular weight` are largely unimportant.  Users should set reasonable numeric values, and keep those value in mind when defining the kinetic rates (below), since they impact the numeric values of the rate parameters.
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:66]: done

File: modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        "O2(aq)": -2.0
+      },
+      ...
+      "logk": [157.892, 144.108, 127.936, 112.88, 97.706, 85.288, 74.75, 65.65]
+    },
+```
+
+Subtracting the second from the first yields [methanogen.cata] and the equilibrium constants.  For instance $160.6192 - 157.892 = 2.7272$ is the first entry.
+
+The advantage of using a kinetic species to represent a microbe is the flexibility, for instance, mortality can be represented as a kinetic rate (see the examples below).  Moreover, in many cases, it is likely that the reactants and products of the catalysed equation are at equilibrium with the aqueous solution, making this approach more suitable than the former approach.  The main disadvantage is the complexity of correctly entering the required information into the database.
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/doc/content/modules/geochemistry/theory/biogeochemistry.md:94]: Yea, i agree, both with the idea of adding a python script, and that it shouldn't be part of this enormous PR.  We'll see how much biogeochem actually gets used

File: modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_death.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+\begin{equation}
+\frac{\mathrm{d}}{\mathrm{d}t}\mathrm{moles} = -k\times \mathrm{moles} = -\frac{k}{\mathrm{molar\ mass}}{\mathrm{mass}} \ .
+\end{equation}
+
+As discussed on the main [biogeochemistry](theory/biogeochemistry.md) page, this can only be implemented in the `geochemistry` module by making the microbe a kinetic species.  Then a [GeochemistryKineticRate](GeochemistryKineticRate.md) `UserObject` may be used to simulate mortality by setting:
+
+- `intrinsic_rate_constant = k/molar_mass`
+- `multiply_by_mass = true`
+- `eta = 0`
+- `direction = DEATH`.
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_death.md:18]: Yes, that is true.  I've now mentioned this explicitly in `biogeochemistry.md`

File: modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_zoning.md (hunk 1 above)
Comment (WilkAndy) [reply modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/bio_zoning.md:147]: Thank you @cpgr .    Your nice words mean a lot to me.

**Comment by WilkAndy** [comment]
@loganharbour - when this is merged, let's move all the images across to large_media.   See #22550

**Comment by WilkAndy** [comment]
@loganharbour - i believe this is ready for a merge now !!

**Comment by WilkAndy** [comment]
Reminding @loganharbour to merge this PR

**Review by loganharbour — CHANGES_REQUESTED** [review]
Please move the new content into large_media (rewrite history for this), otherwise, good for me. We should work on simplifying the utility classes into not having so many function parameters in the future :D

**Comment by WilkAndy** [comment]
@loganharbour , boohoo - i failed with the git somehow.  i hope you can help - see #22611

File: large_media [hunk 6]
```diff
@@ -1 +1 @@
-Subproject commit 88bc8647f9f9954554485d010d7a6311e20fca71
+Subproject commit 50849f069720666d8979e0083bd19dea5dc69e11
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply large_media:1]: this is deliberate !

**Comment by WilkAndy** [comment]
Oh my goodness, this passes!   Can you merge, @loganharbour ?

File: large_media [hunk 7]
```diff
@@ -1 +1 @@
-Subproject commit 889a6e9e9bf06bdbe39220a528fc8f905b9c06df
+Subproject commit 50849f069720666d8979e0083bd19dea5dc69e11
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply large_media]: outdated - i don't need this anymore as someone else has updated large_media

**Comment by rpodgorney** [comment]
@WilkAndy @loganharbour was the move to large media completed?

We need to move this one along ASAP please.......

**Comment by WilkAndy** [comment]
> @WilkAndy @loganharbour was the move to large media completed?
> 
> We need to move this one along ASAP please.......

Yep, confirming the move to large media has been performed

**Review by loganharbour — APPROVED** [review]

**Comment by rpodgorney** [comment]
Thanks @loganharbour

**Comment by WilkAndy** [comment]
Thanks @loganharbour

**Pull request merged**